[MCKIN-2723] Use i18n runtime service to provide translations for HTML templates

pomegranited opened the pull request:
Once the dependent PRs are merged, this change is all that's required to allow this XBlock to use its own translations for text found in HTML templates.

**JIRA tickets**: [MCKIN-7023](https://edx-wiki.atlassian.net/browse/MCKIN-7023), [OSPR-2289](https://openedx.atlassian.net/browse/OSPR-2289)

**Discussions**: [WL-230](https://openedx.atlassian.net/browse/WL-230), [YONK-879](https://openedx.atlassian.net/browse/YONK-879?focusedCommentId=307771&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-307771)

**Dependencies**:

* https://github.com/edx/xblock-utils/pull/48 - merged
* https://github.com/edx/xblock-utils/pull/49 - merged

**Screenshots**: 

**Before** 
these changes, the only text in DnDv2 HTML templates that would appear translated were the text that happened to be translated in edx-platform.
![before 1](https://user-images.githubusercontent.com/7556571/37516056-da821546-295c-11e8-8192-387bd59edf28.png)

**After** this change, DnDv2 template translations like [this one](https://github.com/edx-solutions/xblock-drag-and-drop-v2/blob/master/drag_and_drop_v2/templates/html/drag_and_drop_edit.html#L170-L174) appear in Studio and LMS:
![after 1](https://user-images.githubusercontent.com/7556571/37516104-f8e2c3dc-295c-11e8-89ed-f21fee13b860.png)

**Sandbox URL**: ([3f81e62](https://github.com/edx/edx-platform/pull/17670/commits/3f81e62f749b887cc50c3b9bd4d2b79c589d6a4a))

* LMS: https://pr17670.sandbox.opencraft.hosting/
* Studio: https://studio-pr17670.sandbox.opencraft.hosting/

**Merge deadline**: None

**Testing instructions**:

1. In Studio, create a course and add a Drag and Drop problem (e.g. [sandbox Drag and Drop](https://studio-pr17670.sandbox.opencraft.hosting/container/block-v1:OpenCraft+DnDv2+2018_1+type@vertical+block@a2b13b1603724df192e1c43dd3f1d70d?action=new#))
1. Visit the Studio URL with `/update_lang` appended, and submit `eo` as your selected Language Code.
1. Edit the Drag and Drop problem, and note that both [`blocktrans`](https://github.com/edx-solutions/xblock-drag-and-drop-v2/blob/master/drag_and_drop_v2/templates/html/drag_and_drop_edit.html#L170-L174) and [`trans`](https://github.com/edx-solutions/xblock-drag-and-drop-v2/blob/master/drag_and_drop_v2/templates/html/drag_and_drop_edit.html#L165) template tag strings appear translated.

**Author notes and concerns**:

1. Translations in the Javascript templates are currently not working at all, and will be addressed by a future change.  This issue can be seen in the LMS (student view), and the Studio view, when adding/removing zones.
1. The [process for extracting translations for this XBlock is still quite manual](https://github.com/edx-solutions/xblock-drag-and-drop-v2#i18n-compatibility), and so this change does not attempt to correct the approach, or to provide any missing translations.

**Reviewers**
- [x] @mtyaka 
- [ ] edx-solutions reviewer[s] TBD

edx-webhook commented:
Thanks for the pull request, @pomegranited! I've created [OSPR-2289](https://openedx.atlassian.net/browse/OSPR-2289) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).



If you like, you can add yourself to the [AUTHORS](https://github.com/edx-solutions/xblock-drag-and-drop-v2/blob/master/AUTHORS)      file for this repo, though that isn't required.     Please see the [CONTRIBUTING](https://github.com/edx/edx-platform/blob/master/CONTRIBUTING.rst)     file for more information.

bradenmacdonald commented on drag_and_drop_v2/drag_and_drop_v2.py:
```diff
@@ -342,13 +342,15 @@ def studio_view(self, context):
+        })
 
         fragment = Fragment()
         fragment.add_content(loader.render_template('/templates/html/drag_and_drop_edit.html', context))
```
Can we change from `render_template` to `render_django_template` while we're at it? The former is deprecated since it's ambiguous in an XBlock context.

pomegranited replied:
Fixed with 4187566.

mtyaka commented:
@pomegranited Currently adding a new DnDv2 block to a course is broken, which is why tests on https://github.com/edx/edx-platform/pull/17670 are failing.

This does not seem to be related to this PR, but rather [a recent change we made to this xblock](https://github.com/open-craft/xblock-drag-and-drop-v2/commit/a62d65e20702f5923c9b9c13dc813d33bd96b3e1#diff-cef30416b48c3e5d4dc05c6922a2729bL719). Apparently the line that was removed is still needed and the comment mentioning Dogwood was wrong. Can you please add it back in?

Other than that I tested this and it seems to be working great, but some translations are missing (see screenshots below). It seems that most of the missing translations are from [js_templates.html](https://github.com/open-craft/xblock-drag-and-drop-v2/blob/c9dbe044947005130d508f815a1bf02a9cc55adf/drag_and_drop_v2/templates/html/js_templates.html), which for some reasons uses `{{ i18n ... }}` instead of `{{ trans ... }}`. Maybe we only need to replace that.

I also noticed that the `mo` file in the [eo translations folder](https://github.com/open-craft/xblock-drag-and-drop-v2/tree/c9dbe044947005130d508f815a1bf02a9cc55adf/drag_and_drop_v2/translations/eo/LC_MESSAGES) was updated 4 days ago, but the `po` file has last been updated one year ago. Some missing translations could be because of outdated translation files.

pomegranited commented:
@mtyaka 
> Currently adding a new DnDv2 block to a course is broken, which is why tests on edx/edx-platform#17670 are failing.

Oo, thank you for figuring out why those tests were failing!  Re-added that line with 9bf377e.

> It seems that most of the missing translations are from [js_templates.html](https://github.com/open-craft/xblock-drag-and-drop-v2/blob/c9dbe044947005130d508f815a1bf02a9cc55adf/drag_and_drop_v2/templates/html/js_templates.html), which for some reasons uses `{{ i18n ... }}` instead of `{% trans ... %}`. Maybe we only need to replace that.

The js_templates.html are JS Handlebars templates, and they were supposed to use this [`i18n` service](https://github.com/edx-solutions/xblock-drag-and-drop-v2/blob/dd91427874637b8abad7fdeacaf45dfb50afaee1/drag_and_drop_v2/public/js/drag_and_drop_edit.js#L9), which says it was hooked up to the client runtime `gettext` method.  This wasn't working for me at all though, and I don't know if it ever was.  Check out our [demo Studio](https://studio-demo.opencraft.hosting/container/block-v1:edX+DemoX+Demo_Course+type@vertical+block@47dbd5f836544e61877a483c0b75606c#) to see.

So I've submitted deea454, which converts the [`js_templates.html`](https://github.com/edx-solutions/xblock-drag-and-drop-v2/pull/154/commits/deea454bca909e50e25b0585cc97b9645ab3ebe7#diff-07b1e1746c42caea36bf774a7ddda805) file to a proper Django template, so we can use this new `i18n` service.  It's not pretty, but it works 😄 

> I also noticed that the mo file in the [eo translations folder](https://github.com/open-craft/xblock-drag-and-drop-v2/tree/c9dbe044947005130d508f815a1bf02a9cc55adf/drag_and_drop_v2/translations/eo/LC_MESSAGES) was updated 4 days ago, but the po file has last been updated one year ago. Some missing translations could be because of outdated translation files.

I updated that with this PR to add the compiled `.mo` file for the `eo` test language, since that file was missing from this repo.  I haven't touched the actual translation strings in the `.po` though, so there are likely to still be some missing.

pomegranited commented:
FYI:  I've re-added the "WIP" to the title so we don't forget to merge and tag https://github.com/edx/xblock-utils/pull/48 first.

bradenmacdonald commented on drag_and_drop_v2/public/js/drag_and_drop.js:
```diff
@@ -716,6 +716,7 @@ function DragAndDropBlock(runtime, element, configuration) {
     var renderView = DragAndDropTemplates(configuration);
 
     var $element = $(element);
+    element = $element[0];  // Works around this Studio bug: https://github.com/edx/edx-platform/pull/11433
```
I think that restoring this line is an alternative to https://github.com/edx-solutions/xblock-drag-and-drop-v2/pull/153 . We should pick one approach.

CC @ciuin

pomegranited replied:
Not fussed here -- @mtyaka which method would you prefer?

mtyaka replied:
I slightly prefer the restored approach, but am fine with either.

I still think the root cause is a bug in edx-platform. It sometimes initializes the xblock with a jquery object and sometimes with a plain DOM element.

pomegranited replied:
Ok, since @ciuin's https://github.com/edx-solutions/xblock-drag-and-drop-v2/pull/153 has already been approved, let's go with that.

bradenmacdonald replied:
> I still think the root cause is a bug in edx-platform. It sometimes initializes the xblock with a jquery object and sometimes with a plain DOM element.

It is, but unfortunately [when we've tried to fix it at the platform level](https://github.com/edx/edx-platform/pull/11433), [new things broke](https://github.com/edx/edx-platform/pull/11674).

mtyaka commented:
Thanks @pomegranited!

> So I've submitted deea454, which converts the js_templates.html file to a proper Django template, so we can use this new i18n service. It's not pretty, but it works 😄

Sorry about that - I did not realize the strings in js_templates.html were supposed to be translated by JS and not by django. Since it's JS translations which do not work in xblocks at all as far as I can tell, and we'll probably have to fix that anyway, it might make sense to revert the changes until we fix JS translations properly.

mtyaka commented:
:+1:

- [x] I tested this: I verified that this patch, together with the dependencies correctly translates the studio edit interface. The student view is not translated, because the student view is not using django templates. The student view interface builds the UI entirely via JS, and JS translations are currently not working.
- [x] I read through the code
- [ ] ~~I checked for accessibility issues~~ n/a
- [ ] ~~Includes documentation~~ n/a

pomegranited commented:
Thanks @mtyaka !  I reverted the js_template.html translation change, but kept the commit in a separate branch for future reference: [c9dbe04](https://github.com/open-craft/xblock-drag-and-drop-v2/commit/c9dbe044947005130d508f815a1bf02a9cc55adf).  Also updated the PR description to explain about the JS translation issue.

pomegranited commented:
Squashed 49cf04bdc928c0c4b0f7e0c18b0bd12dea9558af...9bf31fedc0460a977f09d5b270e6eae53e485f8b into 4aeb0f755d858b11115846f36b92aec090d51e1b to make it easier to use this change on the client's fork.

edx-webhook commented:
Thanks for the pull request, @pomegranited! I've created [OSPR-2343](https://openedx.atlassian.net/browse/OSPR-2343) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).



If you like, you can add yourself to the [AUTHORS](https://github.com/edx-solutions/xblock-drag-and-drop-v2/blob/master/AUTHORS)      file for this repo, though that isn't required.     Please see the [CONTRIBUTING](https://github.com/edx/edx-platform/blob/master/CONTRIBUTING.rst)     file for more information.

pomegranited commented:
@sanfordstudent All our dependent PRs have merged, so this PR is ready for review.  (Yay!)
CC @mduboseedx

pomegranited commented:
Thank you @sanfordstudent !  Merging now.
